python/ray/tests/aws/test_autoscaler_aws.py

```diff
@@ -164,6 +166,122 @@ def test_fills_out_amis(iam_client_stub, ec2_client_stub):
+    config.update(head_and_worker_kludge)
+
+    # Generate stubs
+    stubs.configure_iam_role_default(iam_client_stub)
+    stubs.configure_key_pair_default(ec2_client_stub)
+
+    # Only one of these (the one specified in head_node / worker_nodes)
+    # is in the correct vpc.
+    # This list of subnets is generated by the ec2.subnets.all() call
+    # and then ignored, since head_node and worker_nodes already specify
+    # subnet_ids.
+    stubs.describe_a_thousand_subnets_in_different_vpcs(ec2_client_stub)
+
+    # The rest of the stubbing logic is copied from
+    # test_create_sg_with_custom_inbound_rules_and_name.
+
+    # expect to describe the head subnet ID
+    stubs.describe_subnets_echo(ec2_client_stub, DEFAULT_SUBNET)
+    # given no existing security groups within the VPC...
+    stubs.describe_no_security_groups(ec2_client_stub)
+    # expect to create a security group on the head node VPC
+    stubs.create_sg_echo(ec2_client_stub, DEFAULT_SG_WITH_NAME)
+    # expect new head security group details to be retrieved after creation
+    stubs.describe_sgs_on_vpc(
+        ec2_client_stub,
+        [DEFAULT_SUBNET["VpcId"]],
+        [DEFAULT_SG_WITH_NAME],
+    )
+
+    # given custom existing default head security group inbound rules...
+    # expect to authorize both default and custom inbound rules
+    stubs.authorize_sg_ingress(
+        ec2_client_stub,
+        DEFAULT_SG_WITH_NAME_AND_RULES,
+    )
+
+    # given the prior modification to the head security group...
+    # expect the next read of a head security group property to reload it
+    stubs.describe_sg_echo(ec2_client_stub, DEFAULT_SG_WITH_NAME_AND_RULES)
+
+    _get_vpc_id_or_die.cache_clear()
+
+    # given our mocks and the config as input...
+    # expect the config to be validated and bootstrapped successfully
+    bootstrapped_config = helpers.bootstrap_aws_config(config)
+
+    # expect the bootstrapped config to have the custom security group...
+    # name and in bound rules
+    assert bootstrapped_config["provider"]["security_group"][
+        "GroupName"] == DEFAULT_SG_WITH_NAME_AND_RULES["GroupName"]
+    assert config["provider"]["security_group"][
+        "IpPermissions"] == CUSTOM_IN_BOUND_RULES
+
+    # Confirming boostrap config does not currently touch available node types.
+    assert bootstrapped_config["available_node_types"] == bootstrapped_config[
+        "available_node_types"]
+
+    # Confirming head and worker subnet_ids are untouched
+    assert bootstrapped_config["head_node"]["SubnetIds"] ==\
+        config["head_node"]["SubnetIds"] ==\
+        config["worker_nodes"]["SubnetIds"] ==\
+        bootstrapped_config["worker_nodes"]["SubnetIds"]
+
+    # Confirming correct security group got filled for head and workers
+    sg_id = DEFAULT_SG["GroupId"]
+    assert bootstrapped_config["head_node"]["SecurityGroupIds"] == [sg_id]
+    assert bootstrapped_config["worker_nodes"]["SecurityGroupIds"] == [sg_id]
```

Comment:
Huh, interesting - are we still relying on deprecated `head_node` and `worker_node_type` to propagate SecurityGroupIds?

Reply:
Yep -- this also happens in `aws/config.py`.

Reply:
I see. Tangential to this PR, what other behaviour do we need to update in aws/config.py for available_node_types? Is it just subnets, or?